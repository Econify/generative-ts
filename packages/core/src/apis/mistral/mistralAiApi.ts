import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/Either";

import type { ModelApi, ModelRequestOptions } from "../../typeDefs";

import { Template } from "../../utils/template";

import type { FewShotRequestOptions } from "../_shared_interfaces/fewShot";

const templateSource = `{
  "model": "<%= modelId %>",
  "messages": [
    <% if (typeof system !== 'undefined') { %>
    {
      "role": "system",
      "content": "<%= system %>"
    },
    <% } %>
    <% (typeof examplePairs !== 'undefined' ? examplePairs : []).forEach(pair => { %>
    {
      "role": "user",
      "content": "<%= pair.user %>"
    },
    {
      "role": "assistant",
      "content": "<%= pair.assistant %>"
    },
    <% }) %>
    <% (typeof messages !== 'undefined' ? messages : []).forEach(message => { %>
    {
      "role": "<%= message.role %>",
      "content": "<%= message.content %>"
    },
    <% }) %>
    {
      "role": "user",
      "content": "<%= prompt %>"
    }
  ]
  <% if (typeof temperature !== 'undefined') { %>, "temperature": <%= temperature %><% } %>
  <% if (typeof top_p !== 'undefined') { %>, "top_p": <%= top_p %><% } %>
  <% if (typeof max_tokens !== 'undefined') { %>, "max_tokens": <%= max_tokens %><% } %>
  <% if (typeof stream !== 'undefined') { %>, "stream": <%= stream %><% } %>
  <% if (typeof safe_prompt !== 'undefined') { %>, "safe_prompt": <%= safe_prompt %><% } %>
  <% if (typeof random_seed !== 'undefined') { %>, "seed": <%= seed %><% } %>
}`;

export interface MistralAiApiOptions
  extends FewShotRequestOptions,
    ModelRequestOptions {
  messages?: {
    role: "user" | "assistant" | "system";
    content: string;
  }[];
  temperature?: number;
  top_p?: number;
  max_tokens?: number;
  stream?: boolean;
  safe_prompt?: boolean;
  random_seed?: number;
}

export const MistralAiApiTemplate = new Template<MistralAiApiOptions>(
  templateSource,
);

const MistralAiApiResponseCodec = t.type({
  id: t.string,
});

export type MistralAiApiResponse = TypeOf<typeof MistralAiApiResponseCodec>;

export function isMistralAiApiResponse(
  response: unknown,
): response is MistralAiApiResponse {
  return !isLeft(MistralAiApiResponseCodec.decode(response));
}

export const MistralAiApi: ModelApi<MistralAiApiOptions, MistralAiApiResponse> =
  {
    requestTemplate: MistralAiApiTemplate,
    responseGuard: isMistralAiApiResponse,
  };
