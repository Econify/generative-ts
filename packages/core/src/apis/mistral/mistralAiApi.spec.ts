import { MistralAiApi, MistralAiOptions } from "./mistralAiApi";

function render(context: Omit<MistralAiOptions, "modelId">) {
  try {
    const rendered = MistralAiApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("MistralAiApi:", () => {
  /**
   * FewShotRequestOptions ($prompt, examplePairs, system):
   */

  test("$prompt", () => {
    const rendered = render({
      $prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, system", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs, system", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * "Native" few shot options (messages):
   */

  test("$prompt, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
        {
          role: "user",
          content: "mock-user-text",
        },
        {
          role: "assistant",
          content: "mock-assistant-text-2",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, system, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      system: "mock-system-text",
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
        {
          role: "user",
          content: "mock-user-text",
        },
        {
          role: "assistant",
          content: "mock-assistant-text-2",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs, system, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system: "mock-system-text",
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */

  test("all options", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
      ],
      system: "mock-system-text",
      temperature: 0.7,
      top_p: 0.9,
      max_tokens: 1000,
      stream: true,
      safe_prompt: true,
      random_seed: 1234,
    });

    expect(rendered).toMatchSnapshot();
  });
});
