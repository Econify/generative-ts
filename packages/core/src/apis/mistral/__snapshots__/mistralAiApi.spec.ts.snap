// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MistralAiApi: $prompt 1`] = `
{
  "messages": [
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt 2`] = `
{
  "messages": [
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt, examplePairs 1`] = `
{
  "messages": [
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-user-msg-2",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-2",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt, examplePairs, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt, examplePairs, system 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt, examplePairs, system, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-user-text",
      "role": "user",
    },
    {
      "content": "mock-assistant-text-2",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt, system 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: $prompt, system, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-user-text",
      "role": "user",
    },
    {
      "content": "mock-assistant-text-2",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`MistralAiApi: all options 1`] = `
{
  "max_tokens": 1000,
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
  "random_seed": 1234,
  "safe_prompt": true,
  "stream": true,
  "temperature": 0.7,
  "top_p": 0.9,
}
`;
